gen/service.go

```diff
@@ -0,0 +1,225 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package gen
+
+import (
+	"bytes"
+	"fmt"
+	"go/token"
+	"strings"
+
+	"github.com/thriftrw/thriftrw-go/compile"
+)
+
+// Service generates code for the given service.
+//
+// Returns a map from file name to contents for that file. The file names are
+// relative to the package directory for the service.
+func Service(g Generator, s *compile.ServiceSpec) (map[string]*bytes.Buffer, error) {
+	files := make(map[string]*bytes.Buffer)
+
+	// TODO inherited service functions
+
+	for _, functionName := range sortStringKeys(s.Functions) {
+		err := ServiceFunction(g, s, s.Functions[functionName])
+		if err != nil {
+			return nil, fmt.Errorf(
+				"could not generate types for %s.%s: %v",
+				s.Name, functionName, err)
+		}
+
+		buff := new(bytes.Buffer)
+		if err := g.Write(buff, token.NewFileSet()); err != nil {
+			return nil, fmt.Errorf("could not write %s.%s: %v", s.Name, functionName, err)
+		}
+
+		// TODO check for conflicts
+		files[strings.ToLower(functionName)+".go"] = buff
+	}
+
+	return files, nil
+}
+
+// ServiceFunction generates code for the given function of the given service.
+func ServiceFunction(g Generator, s *compile.ServiceSpec, f *compile.FunctionSpec) error {
+	if f.OneWay {
+		return fmt.Errorf(
+			"oneway functions are not yet supported: %s.%s is oneway",
+			s.Name, f.Name)
+	}
+
+	argsGen := fieldGroupGenerator{
+		Name:   goCase(f.Name) + "Args",
+		Fields: compile.FieldGroup(f.ArgsSpec),
+	}
+	if err := argsGen.Generate(g); err != nil {
+		return wrapGenerateError(fmt.Sprintf("%s.%s", s.Name, f.Name), err)
+	}
+
+	resultFields := make(compile.FieldGroup, 0)
+	if result := f.ResultSpec; result != nil {
+		if result.ReturnType != nil {
+			resultFields = append(resultFields, &compile.FieldSpec{
+				ID:   0,
+				Name: "success",
+				Type: result.ReturnType,
+			})
+		}
+
+		resultFields = append(resultFields, result.Exceptions...)
+	}
+	resultGen := fieldGroupGenerator{
+		Name:   goCase(f.Name) + "Result",
+		Fields: resultFields,
+	}
+	if err := resultGen.Generate(g); err != nil {
+		return wrapGenerateError(fmt.Sprintf("%s.%s", s.Name, f.Name), err)
+	}
+
+	if err := functionHelper(g, f); err != nil {
+		return wrapGenerateError(fmt.Sprintf("%s.%s", s.Name, f.Name), err)
+	}
+
+	// TODO(abg): If we receive unknown exceptions over the wire, we need to
+	// throw a generic error.
+	return nil
+}
+
+func functionHelper(g Generator, f *compile.FunctionSpec) error {
+	return g.DeclareFromTemplate(
+		`
+		<$name := goCase .Name>
+
+		var <$name> = struct{
+			IsException func(error) bool
+
+			Args func(
+				<range .ArgsSpec>
+					<if .Required>
+						<.Name> <typeReference .Type>,
```

Comment:
if the `Name` clashes with a go keyword, this will fail to compile. the full list is at 
https://golang.org/ref/spec#Keywords